src/Scenario/Processor.php

```diff
@@ -64,43 +63,55 @@ public function execute(array $scenarios): void
             implode(', ', $scenarios),
             $this->packageManager->getPrettyInfo()
         ));
+        $steps = [];
 
         try {
             $this->mergedScenarios = $this->merger->merge($scenarios);
             $steps = $this->mergedScenarios['steps'];
+        } catch (Throwable $exception) {
+            $this->handleException(
+                $exception,
+                sprintf('Unhandled error: %s', $exception->getMessage())
+            );
+        }
 
+        try {
             array_walk($steps, function (StepInterface $step, string $name) {
                 $this->logger->debug('Running step: ' . $name);
 
                 $step->execute();
 
                 $this->logger->debug(sprintf('Step "%s" finished', $name));
             });
-        } catch (StepException $stepException) {
-            try {
-                $actions = $this->mergedScenarios['actions'];
-
-                array_walk($actions, function (ActionInterface $action, string $name) {
-                    $this->logger->debug('Running on fail action: ' . $name);
-
-                    $action->execute();
-
-                    $this->logger->debug(sprintf('On fail action "%s" finished', $name));
-                });
-            } catch (ActionException $actionException) {
-                $this->logger->error($actionException->getMessage());
-            }
+        } catch (Throwable $stepException) {
```

Comment:
Looks like here should be `StepException` instead of `Throwable`, as `Step` can throw only `StepException`

Reply:
you can change this method to something like this with one Try/Catch block
```php
    public function execute(array $scenarios): void
    {
        try {
            $this->logger->info(sprintf(
                'Starting scenario(s): %s %s',
                implode(', ', $scenarios),
                $this->packageManager->getPrettyInfo()
            ));

            $this->mergedScenarios = $this->merger->merge($scenarios);
            $steps = $this->mergedScenarios['steps'];

            array_walk($steps, function (StepInterface $step, string $name) {
                $this->logger->debug('Running step: ' . $name);

                $step->execute();

                $this->logger->debug(sprintf('Step "%s" finished', $name));
            });

            $this->logger->info('Scenario(s) finished');
        } catch (StepException $stepException) {
            try {
                $actions = $this->mergedScenarios['actions'];

                array_walk($actions, function (ActionInterface $action, string $name) {
                    $this->logger->debug('Running on fail action: ' . $name);

                    $action->execute();

                    $this->logger->debug(sprintf('On fail action "%s" finished', $name));
                });
            } catch (ActionException $actionException) {
                $this->logger->error($actionException->getMessage());
            }
            $this->handleException($stepException);
        } catch (Throwable $exception) {
            $this->handleException(
                $exception,
                sprintf('Unhandled error: [%d] %s', $exception->getCode(), $exception->getMessage())
            );
        }
    }
```

Reply:
no, we should catch Throwable, otherwise there may be a case when we did not set flag that deploy was failed and post-deploy phase will be executed.